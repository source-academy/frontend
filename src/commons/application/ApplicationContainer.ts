import { connect, MapDispatchToProps, MapStateToProps } from 'react-redux';
import { withRouter } from 'react-router';
import { bindActionCreators, Dispatch } from 'redux';
import { createCourse } from 'src/features/academy/AcademyActions';

import { logOut } from './actions/CommonsActions';
import {
  fetchUserAndCourse,
  loginGitHub,
  logoutGitHub,
  updateCourseResearchAgreement,
  updateLatestViewedCourse
} from './actions/SessionActions';
import Application, { DispatchProps, StateProps } from './Application';
import { OverallState } from './ApplicationTypes';

/**
 * Provides the title of the application for display.
 * An object with the relevant properties must be
 * returned instead of an object of type @type {IApplicationProps},
 * as the routing properties of @type {RouteComponentProps} are
 * provided using the withRouter() method below.
 */
const mapStateToProps: MapStateToProps<StateProps, {}, OverallState> = state => ({
  role: state.session.role,
  name: state.session.name,
  courses: state.session.courses,
  courseId: state.session.courseId,
  courseShortName: state.session.courseShortName,
  enableAchievements: state.session.enableAchievements,
  enableSourcecast: state.session.enableSourcecast,
  assessmentTypes: state.session.assessmentConfigurations?.map(e => e.type),
  agreedToResearch: state.session.agreedToResearch
});

const mapDispatchToProps: MapDispatchToProps<DispatchProps, {}> = (dispatch: Dispatch) =>
  bindActionCreators(
    {
      handleLogOut: logOut,
      handleGitHubLogIn: loginGitHub,
      handleGitHubLogOut: logoutGitHub,
      fetchUserAndCourse: fetchUserAndCourse,
      updateLatestViewedCourse: updateLatestViewedCourse,
      handleCreateCourse: createCourse,
      updateCourseResearchAgreement: updateCourseResearchAgreement
    },
    dispatch
  );

const ApplicationContainer = withRouter(connect(mapStateToProps, mapDispatchToProps)(Application));

export default ApplicationContainer;
