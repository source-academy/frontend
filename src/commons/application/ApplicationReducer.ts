import { Action, Reducer } from 'redux';

import { ApplicationState, defaultApplication } from './ApplicationTypes';

export const ApplicationReducer: Reducer<ApplicationState> = (
  state = defaultApplication,
  action: Action
) => {
  return state;
};
