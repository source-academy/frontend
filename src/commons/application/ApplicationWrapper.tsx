import { useMemo } from 'react';
import { useDispatch } from 'react-redux';
import { RouterProvider } from 'react-router';
import { createBrowserRouter } from 'react-router-dom';

import { getFullAcademyRouterConfig, playgroundOnlyRouterConfig } from '../../routes/routerConfig';
import Constants from '../utils/Constants';
import { useSession } from '../utils/Hooks';
import { updateReactRouter } from './actions/CommonsActions';

/**
 * Application wrapper component which figures out which deployment and set of routes to render.
 *
 * There are 3 main types of deployments as follows:
 * 1. Playground-only (stripped-down backendless version of SA - e.g. https://sourceacademy.org)
 * 2. Full Academy (full SA to be deployed and configured with the backend - e.g. https://sourceacademy.nus.edu.sg)
 * 3. Disabled (disabled SA which only allows `staff` and `admin` accounts to log in - e.g. during examinations)
 */
const ApplicationWrapper: React.FC = () => {
  const dispatch = useDispatch();
  const { isLoggedIn, role, name, courseId } = useSession();

  const router = useMemo(() => {
    const routerConfig = Constants.playgroundOnly
      ? playgroundOnlyRouterConfig
      : getFullAcademyRouterConfig({
          name,
          role,
          isLoggedIn,
          courseId
        });

    const r = createBrowserRouter(routerConfig);
    dispatch(updateReactRouter(r));

    return r;
  }, [isLoggedIn, role, name, courseId, dispatch]);

  return <RouterProvider router={router} />;
};

export default ApplicationWrapper;
