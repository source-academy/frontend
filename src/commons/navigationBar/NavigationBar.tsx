import {
  Alignment,
  Button,
  Classes,
  Drawer,
  FocusStyleManager,
  Icon,
  Navbar,
  NavbarDivider,
  NavbarGroup,
  NavbarHeading,
  Position
} from '@blueprintjs/core';
import { IconNames } from '@blueprintjs/icons';
import { BlueprintIcons_16Id } from '@blueprintjs/icons/lib/esm/generated-icons/16px/blueprint-icons-16';
import { Popover2 } from '@blueprintjs/popover2';
import classNames from 'classnames';
import { Location } from 'history';
import { useCallback, useMemo, useState } from 'react';
import { NavLink, Route, Routes, useLocation } from 'react-router-dom';

import Dropdown from '../dropdown/Dropdown';
import NotificationBadge from '../notificationBadge/NotificationBadge';
import { filterNotificationsByType } from '../notificationBadge/NotificationBadgeHelper';
import Constants from '../utils/Constants';
import { useResponsive, useSession } from '../utils/Hooks';
import AcademyNavigationBar, {
  assessmentTypesToNavlinkInfo,
  getAcademyNavbarRightInfo
} from './subcomponents/AcademyNavigationBar';
import NavigationBarLangSelectButton from './subcomponents/NavigationBarLangSelectButton';
import SicpNavigationBar from './subcomponents/SicpNavigationBar';

export type NavbarEntryInfo = {
  to: string;
  icon: BlueprintIcons_16Id;
  text: string;
  disabled?: boolean; // entry is not rendered when disabled
  hasNotifications?: boolean; // whether to render NotificationBadge
  hiddenInBreakpoints?: ('xs' | 'sm' | 'md' | 'lg')[]; // hide text in Blueprint breakpoints
};

type CreateNavlinkFunction = (navbarEntry: NavbarEntryInfo) => React.ReactElement;

const NavigationBar: React.FC = () => {
  const [mobileSideMenuOpen, setMobileSideMenuOpen] = useState(false);
  const { isMobileBreakpoint } = useResponsive();
  const location = useLocation();
  const {
    isLoggedIn,
    isEnrolledInACourse,
    role,
    courseId,
    courseShortName,
    enableAchievements,
    enableSourcecast,
    assessmentConfigurations
  } = useSession();
  const assessmentTypes = useMemo(
    () => assessmentConfigurations?.map(c => c.type),
    [assessmentConfigurations]
  );

  FocusStyleManager.onlyShowFocusOnTabs();

  const createMobileNavlink: CreateNavlinkFunction = useCallback(
    navbarEntry => (
      <NavLink
        to={navbarEntry.to}
        className={({ isActive }) =>
          classNames(
            'NavigationBar__link__mobile',
            Classes.BUTTON,
            Classes.MINIMAL,
            Classes.LARGE,
            {
              [Classes.ACTIVE]: isActive
            }
          )
        }
        onClick={() => setMobileSideMenuOpen(false)}
        key={navbarEntry.text}
      >
        <Icon icon={navbarEntry.icon} />
        <div>{navbarEntry.text}</div>
        {navbarEntry.hasNotifications && (
          <NotificationBadge
            notificationFilter={filterNotificationsByType(navbarEntry.text)}
            disableHover={true}
          />
        )}
      </NavLink>
    ),
    [setMobileSideMenuOpen]
  );

  const wrapWithMobileHamburger = (navlinks: (React.ReactElement | null)[]) => {
    // Don't render drawer when there are 0 navlinks in it
    const nonNullNavlinks = navlinks.filter(e => e !== null);
    const renderDrawer = nonNullNavlinks.length > 0;

    return (
      <NavbarGroup align={Alignment.LEFT}>
        {renderDrawer && (
          <Button
            onClick={() => setMobileSideMenuOpen(!mobileSideMenuOpen)}
            icon={IconNames.MENU}
            large={true}
            minimal={true}
          />
        )}
        <NavLink
          className={classNames('NavigationBar__link', Classes.BUTTON, Classes.MINIMAL)}
          to={
            Constants.playgroundOnly ? '/' : courseId == null ? '/welcome' : `/courses/${courseId}`
          }
        >
          <Icon icon={IconNames.SYMBOL_DIAMOND} />
          <NavbarHeading style={{ paddingBottom: '0px' }}>
            {courseShortName || Constants.sourceAcademyDeploymentName}
          </NavbarHeading>
        </NavLink>
        {renderDrawer && (
          <Drawer
            isOpen={mobileSideMenuOpen}
            position="left"
            onClose={() => setMobileSideMenuOpen(false)}
            title=""
            className={Classes.DARK}
            style={{ overflowY: 'auto' }}
          >
            {navlinks}
          </Drawer>
        )}
      </NavbarGroup>
    );
  };

  const fullAcademyNavbarLeftAssessmentsInfo: NavbarEntryInfo[] = useMemo(
    () =>
      assessmentTypesToNavlinkInfo({
        assessmentTypes,
        courseId,
        isEnrolledInACourse
      }),
    [assessmentTypes, courseId, isEnrolledInACourse]
  );

  const fullAcademyNavbarLeftCommonInfo: NavbarEntryInfo[] = useMemo(() => {
    return [
      {
        to: `/courses/${courseId}/sourcecast`,
        icon: IconNames.MUSIC,
        text: 'Sourcecast',
        disabled: !(isEnrolledInACourse && enableSourcecast)
      },
      {
        to: '/playground',
        icon: IconNames.CODE,
        text: 'Playground',
        disabled: !isEnrolledInACourse
      },
      {
        to: '/githubassessments',
        icon: IconNames.BRIEFCASE,
        text: 'Classroom',
        disabled: !Constants.enableGitHubAssessments
      },
      {
        to: '/sicpjs',
        icon: IconNames.BOOK,
        text: 'SICP JS',
        disabled: !isLoggedIn
      },
      {
        to: `/courses/${courseId}/achievements`,
        icon: IconNames.MOUNTAIN,
        text: 'Achievements',
        disabled: !(isEnrolledInACourse && enableAchievements)
      },
      {
        to: '/stories',
        icon: IconNames.GIT_REPO,
        text: 'Stories',
        // TODO: Enable when stories are implemented
        disabled: true && !isLoggedIn
      }
    ];
  }, [isLoggedIn, isEnrolledInACourse, courseId, enableSourcecast, enableAchievements]);

  const fullAcademyMobileNavbarLeftAdditionalInfo = useMemo(
    () => getAcademyNavbarRightInfo({ isEnrolledInACourse, courseId, role }),
    [isEnrolledInACourse, courseId, role]
  );

  const fullAcademyMobileNavbarLeftInfoWithAssessments: NavbarEntryInfo[] = useMemo(() => {
    return [
      ...fullAcademyNavbarLeftAssessmentsInfo,
      ...fullAcademyNavbarLeftCommonInfo,
      ...fullAcademyMobileNavbarLeftAdditionalInfo
    ];
  }, [
    fullAcademyNavbarLeftAssessmentsInfo,
    fullAcademyNavbarLeftCommonInfo,
    fullAcademyMobileNavbarLeftAdditionalInfo
  ]);

  const renderPlaygroundOnlyNavbarLeftDesktop = () => (
    <NavbarGroup align={Alignment.LEFT}>
      {renderNavlinksFromInfo(playgroundOnlyNavbarLeftInfo, createDesktopNavlink)}
    </NavbarGroup>
  );

  const renderPlaygroundOnlyNavbarLeftMobile = () =>
    wrapWithMobileHamburger(
      renderNavlinksFromInfo(playgroundOnlyNavbarLeftInfo, createMobileNavlink)
    );

  const renderFullAcademyNavbarLeftDesktop = () => {
    const desktopNavbarLeftPopoverContent = (
      <Navbar>
        <NavbarGroup>
          {renderNavlinksFromInfo(
            assessmentTypesToNavlinkInfo({
              assessmentTypes,
              courseId,
              isEnrolledInACourse
            }),
            createDesktopNavlink
          )}
        </NavbarGroup>
      </Navbar>
    );
    const topNavbarNavlinks = [
      '/playground',
      '/sicpjs',
      '/contributors',
      '/githubassessments',
      `/courses/${courseId}/sourcecast`,
      `/courses/${courseId}/achievements`
    ];
    const enableDesktopPopover =
      courseId != null && !!topNavbarNavlinks.find(x => location.pathname.startsWith(x));
    const highlightDesktopLogo = (location: Location) => {
      // Highlight main logo when none of the topmost-blue navbar links are active
      return !topNavbarNavlinks.find(x => location.pathname.startsWith(x));
    };

    return (
      <NavbarGroup align={Alignment.LEFT}>
        <Popover2
          position={Position.BOTTOM_RIGHT}
          interactionKind="hover"
          content={desktopNavbarLeftPopoverContent}
          popoverClassName={'desktop-navbar-popover'}
          disabled={!enableDesktopPopover}
        >
          <NavLink
            className={classNames('NavigationBar__link', Classes.BUTTON, Classes.MINIMAL, {
              [Classes.ACTIVE]: highlightDesktopLogo(location)
            })}
            to={courseId == null ? '/welcome' : `/courses/${courseId}`}
          >
            <Icon icon={IconNames.SYMBOL_DIAMOND} />
            <NavbarHeading style={{ paddingBottom: '0px' }}>
              {courseShortName || Constants.sourceAcademyDeploymentName}
            </NavbarHeading>
          </NavLink>
        </Popover2>
        {renderNavlinksFromInfo(fullAcademyNavbarLeftCommonInfo, createDesktopNavlink)}
      </NavbarGroup>
    );
  };

  const renderFullAcademyNavbarLeftMobile = () =>
    wrapWithMobileHamburger(
      renderNavlinksFromInfo(fullAcademyMobileNavbarLeftInfoWithAssessments, createMobileNavlink)
    );

  const commonNavbarRight = (
    <NavbarGroup align={Alignment.RIGHT}>
      {location.pathname.startsWith('/playground') && <NavigationBarLangSelectButton />}
      <NavLink
        className={({ isActive }) =>
          classNames('NavigationBar__link', Classes.BUTTON, Classes.MINIMAL, {
            [Classes.ACTIVE]: isActive
          })
        }
        to="/contributors"
      >
        <Icon icon={IconNames.HEART} />
      </NavLink>

      <div className="visible-xs">
        <NavbarDivider className="thin-divider" />
      </div>
      <div className="hidden-xs">
        <NavbarDivider className="default-divider" />
      </div>

      <Dropdown />
    </NavbarGroup>
  );

  return (
    <>
      <Navbar className={classNames('NavigationBar', 'primary-navbar', Classes.DARK)}>
        {Constants.playgroundOnly
          ? isMobileBreakpoint
            ? renderPlaygroundOnlyNavbarLeftMobile()
            : renderPlaygroundOnlyNavbarLeftDesktop()
          : isMobileBreakpoint
          ? renderFullAcademyNavbarLeftMobile()
          : renderFullAcademyNavbarLeftDesktop()}
        {commonNavbarRight}
      </Navbar>

      <Routes>
        <Route path="/playground" element={null} />
        <Route path="/githubassessments/*" element={null} />
        <Route path="/contributors" element={null} />
        <Route path="/courses/:courseId/sourcecast" element={null} />
        <Route path="/courses/:courseId/achievements" element={null} />
        <Route path="/sicpjs/:section?" element={<SicpNavigationBar />} />
        <Route
          path="*"
          element={
            !Constants.playgroundOnly && isEnrolledInACourse && !isMobileBreakpoint ? (
              <AcademyNavigationBar assessmentTypes={assessmentTypes} />
            ) : null
          }
        />
      </Routes>
    </>
  );
};

const playgroundOnlyNavbarLeftInfo: NavbarEntryInfo[] = [
  {
    to: '/playground',
    icon: IconNames.CODE,
    text: 'Playground'
  },
  {
    to: '/githubassessments',
    icon: IconNames.BRIEFCASE,
    text: 'Classroom',
    disabled: !Constants.enableGitHubAssessments
  },
  {
    to: '/sicpjs',
    icon: IconNames.BOOK,
    text: 'SICP JS'
  },
  {
    to: '/stories',
    icon: IconNames.GIT_REPO,
    text: 'Stories',
    // TODO: Enable when stories are implemented
    disabled: true
  }
];

export const renderNavlinksFromInfo = (
  navbarEntries: NavbarEntryInfo[],
  createNavlink: CreateNavlinkFunction
): (React.ReactElement | null)[] =>
  navbarEntries.map(entry => {
    if (entry.disabled) {
      return null;
    }

    return createNavlink(entry);
  });

export const createDesktopNavlink: CreateNavlinkFunction = navbarEntry => (
  <NavLink
    className={({ isActive }) =>
      classNames('NavigationBar__link', Classes.BUTTON, Classes.MINIMAL, {
        [Classes.ACTIVE]: isActive
      })
    }
    to={navbarEntry.to}
    key={navbarEntry.text}
    title={navbarEntry.text}
  >
    <Icon icon={navbarEntry.icon} />
    <div
      className={classNames(
        'navbar-button-text',
        navbarEntry.hiddenInBreakpoints?.map(bp => `hidden-${bp}`)
      )}
    >
      {navbarEntry.text}
    </div>
    {navbarEntry.hasNotifications && (
      <NotificationBadge
        notificationFilter={filterNotificationsByType(navbarEntry.text)}
        disableHover={true}
      />
    )}
  </NavLink>
);

export default NavigationBar;
