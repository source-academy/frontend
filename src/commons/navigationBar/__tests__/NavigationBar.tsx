import { shallow } from 'enzyme';

import { Role } from '../../application/ApplicationTypes';
import NavigationBar from '../NavigationBar';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: () => ({
    pathname: 'localhost:8000/courses/1/game'
  })
}));

test('NavigationBar renders "Not logged in" correctly', () => {
  const props = {
    handleLogOut: () => {},
    handleGitHubLogIn: () => {},
    handleGitHubLogOut: () => {},
    updateLatestViewedCourse: () => {},
    handleCreateCourse: () => {},
    courses: [],
    assessmentTypes: []
  };
  const tree = shallow(<NavigationBar {...props} />);
  expect(tree.debug()).toMatchSnapshot();
});

test('NavigationBar renders correctly for student with course', () => {
  const props = {
    handleLogOut: () => {},
    handleGitHubLogIn: () => {},
    handleGitHubLogOut: () => {},
    updateLatestViewedCourse: () => {},
    handleCreateCourse: () => {},
    courses: [
      {
        courseId: 1,
        courseName: 'CS1101S Programming Methodology (AY20/21 Sem 1)',
        courseShortName: 'CS1101S',
        role: Role.Admin,
        viewable: true
      }
    ],
    courseId: 1,
    courseShortName: 'CS1101S',
    enableAchievements: true,
    enableSourcecast: true,
    role: Role.Student,
    name: 'Bob',
    assessmentTypes: ['Missions', 'Quests', 'Paths', 'Contests', 'Others']
  };
  const tree = shallow(<NavigationBar {...props} />);
  expect(tree.debug()).toMatchSnapshot();
});

test('NavigationBar renders correctly for student without course', () => {
  const props = {
    handleLogOut: () => {},
    handleGitHubLogIn: () => {},
    handleGitHubLogOut: () => {},
    updateLatestViewedCourse: () => {},
    handleCreateCourse: () => {},
    courses: [],
    name: 'Bob',
    assessmentTypes: []
  };
  const tree = shallow(<NavigationBar {...props} />);
  expect(tree.debug()).toMatchSnapshot();
});
