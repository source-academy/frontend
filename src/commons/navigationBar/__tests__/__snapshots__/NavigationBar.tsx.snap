// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationBar renders "Not logged in" correctly 1`] = `
"<Fragment>
  <Blueprint4.Navbar className=\\"NavigationBar primary-navbar bp4-dark\\">
    <Blueprint4.NavbarGroup align=\\"left\\">
      <NavLink activeClassName=\\"bp4-active\\" className=\\"NavigationBar__link bp4-button bp4-minimal\\" to=\\"/welcome\\" isActive={[Function: matchExceptSourcecastAchievement]}>
        <Blueprint4.Icon icon=\\"symbol-diamond\\" />
        <Blueprint4.NavbarHeading style={{...}}>
          Source Academy @ NUS
        </Blueprint4.NavbarHeading>
      </NavLink>
      <NavLink activeClassName=\\"bp4-active\\" className=\\"NavigationBar__link bp4-button bp4-minimal\\" to=\\"/githubassessments\\">
        <Blueprint4.Icon icon=\\"briefcase\\" />
        <div className=\\"navbar-button-text\\">
          Classroom
        </div>
      </NavLink>
    </Blueprint4.NavbarGroup>
    <Blueprint4.NavbarGroup align=\\"right\\">
      <NavLink activeClassName=\\"bp4-active\\" className=\\"NavigationBar__link bp4-button bp4-minimal\\" to=\\"/contributors\\">
        <Blueprint4.Icon icon=\\"heart\\" />
      </NavLink>
      <div className=\\"visible-xs\\">
        <Blueprint4.NavbarDivider className=\\"thin-divider\\" />
      </div>
      <div className=\\"hidden-xs\\">
        <Blueprint4.NavbarDivider className=\\"default-divider\\" />
      </div>
      <Dropdown handleLogOut={[Function: handleLogOut]} handleCreateCourse={[Function: handleCreateCourse]} courses={{...}} courseId={[undefined]} name={[undefined]} />
    </Blueprint4.NavbarGroup>
  </Blueprint4.Navbar>
  <Switch>
    <Route path=\\"/playground\\" />
    <Route path=\\"/githubassessments\\" />
    <Route path=\\"/contributors\\" />
    <Route path=\\"/courses/:courseId/sourcecast\\" />
    <Route path=\\"/courses/:courseId/achievements\\" />
    <Route path=\\"/sicpjs/:section?\\">
      <SicpNavigationBar />
    </Route>
    <Route />
  </Switch>
</Fragment>"
`;

exports[`NavigationBar renders correctly for student with course 1`] = `
"<Fragment>
  <Blueprint4.Navbar className=\\"NavigationBar primary-navbar bp4-dark\\">
    <Blueprint4.NavbarGroup align=\\"left\\">
      <NavLink activeClassName=\\"bp4-active\\" className=\\"NavigationBar__link bp4-button bp4-minimal\\" to=\\"/courses/1\\" isActive={[Function: matchExceptSourcecastAchievement]}>
        <Blueprint4.Icon icon=\\"symbol-diamond\\" />
        <Blueprint4.NavbarHeading style={{...}}>
          CS1101S
        </Blueprint4.NavbarHeading>
      </NavLink>
      <NavLink activeClassName=\\"bp4-active\\" className=\\"NavigationBar__link bp4-button bp4-minimal\\" to=\\"/courses/1/sourcecast\\">
        <Blueprint4.Icon icon=\\"music\\" />
        <div className=\\"navbar-button-text\\">
          Sourcecast
        </div>
      </NavLink>
      <NavLink activeClassName=\\"bp4-active\\" className=\\"NavigationBar__link bp4-button bp4-minimal\\" to=\\"/playground\\">
        <Blueprint4.Icon icon=\\"code\\" />
        <div className=\\"navbar-button-text\\">
          Playground
        </div>
      </NavLink>
      <NavLink activeClassName=\\"bp4-active\\" className=\\"NavigationBar__link bp4-button bp4-minimal\\" to=\\"/githubassessments\\">
        <Blueprint4.Icon icon=\\"briefcase\\" />
        <div className=\\"navbar-button-text\\">
          Classroom
        </div>
      </NavLink>
      <NavLink activeClassName=\\"bp4-active\\" className=\\"NavigationBar__link bp4-button bp4-minimal\\" to=\\"/sicpjs/\\">
        <Blueprint4.Icon icon=\\"book\\" />
        <div className=\\"navbar-button-text\\">
          SICP JS
        </div>
      </NavLink>
      <NavLink activeClassName=\\"bp4-active\\" className=\\"NavigationBar__link bp4-button bp4-minimal\\" to=\\"/courses/1/achievements\\">
        <Blueprint4.Icon icon=\\"mountain\\" />
        <div className=\\"navbar-button-text\\">
          Achievements
        </div>
      </NavLink>
    </Blueprint4.NavbarGroup>
    <Blueprint4.NavbarGroup align=\\"right\\">
      <NavLink activeClassName=\\"bp4-active\\" className=\\"NavigationBar__link bp4-button bp4-minimal\\" to=\\"/contributors\\">
        <Blueprint4.Icon icon=\\"heart\\" />
      </NavLink>
      <div className=\\"visible-xs\\">
        <Blueprint4.NavbarDivider className=\\"thin-divider\\" />
      </div>
      <div className=\\"hidden-xs\\">
        <Blueprint4.NavbarDivider className=\\"default-divider\\" />
      </div>
      <Dropdown handleLogOut={[Function: handleLogOut]} handleCreateCourse={[Function: handleCreateCourse]} courses={{...}} courseId={1} name=\\"Bob\\" />
    </Blueprint4.NavbarGroup>
  </Blueprint4.Navbar>
  <Switch>
    <Route path=\\"/playground\\" />
    <Route path=\\"/githubassessments\\" />
    <Route path=\\"/contributors\\" />
    <Route path=\\"/courses/:courseId/sourcecast\\" />
    <Route path=\\"/courses/:courseId/achievements\\" />
    <Route path=\\"/sicpjs/:section?\\">
      <SicpNavigationBar />
    </Route>
    <Route>
      <AcademyNavigationBar role=\\"student\\" assessmentTypes={{...}} courseId={1} />
    </Route>
  </Switch>
</Fragment>"
`;

exports[`NavigationBar renders correctly for student without course 1`] = `
"<Fragment>
  <Blueprint4.Navbar className=\\"NavigationBar primary-navbar bp4-dark\\">
    <Blueprint4.NavbarGroup align=\\"left\\">
      <NavLink activeClassName=\\"bp4-active\\" className=\\"NavigationBar__link bp4-button bp4-minimal\\" to=\\"/welcome\\" isActive={[Function: matchExceptSourcecastAchievement]}>
        <Blueprint4.Icon icon=\\"symbol-diamond\\" />
        <Blueprint4.NavbarHeading style={{...}}>
          Source Academy @ NUS
        </Blueprint4.NavbarHeading>
      </NavLink>
      <NavLink activeClassName=\\"bp4-active\\" className=\\"NavigationBar__link bp4-button bp4-minimal\\" to=\\"/githubassessments\\">
        <Blueprint4.Icon icon=\\"briefcase\\" />
        <div className=\\"navbar-button-text\\">
          Classroom
        </div>
      </NavLink>
      <NavLink activeClassName=\\"bp4-active\\" className=\\"NavigationBar__link bp4-button bp4-minimal\\" to=\\"/sicpjs/\\">
        <Blueprint4.Icon icon=\\"book\\" />
        <div className=\\"navbar-button-text\\">
          SICP JS
        </div>
      </NavLink>
    </Blueprint4.NavbarGroup>
    <Blueprint4.NavbarGroup align=\\"right\\">
      <NavLink activeClassName=\\"bp4-active\\" className=\\"NavigationBar__link bp4-button bp4-minimal\\" to=\\"/contributors\\">
        <Blueprint4.Icon icon=\\"heart\\" />
      </NavLink>
      <div className=\\"visible-xs\\">
        <Blueprint4.NavbarDivider className=\\"thin-divider\\" />
      </div>
      <div className=\\"hidden-xs\\">
        <Blueprint4.NavbarDivider className=\\"default-divider\\" />
      </div>
      <Dropdown handleLogOut={[Function: handleLogOut]} handleCreateCourse={[Function: handleCreateCourse]} courses={{...}} courseId={[undefined]} name=\\"Bob\\" />
    </Blueprint4.NavbarGroup>
  </Blueprint4.Navbar>
  <Switch>
    <Route path=\\"/playground\\" />
    <Route path=\\"/githubassessments\\" />
    <Route path=\\"/contributors\\" />
    <Route path=\\"/courses/:courseId/sourcecast\\" />
    <Route path=\\"/courses/:courseId/achievements\\" />
    <Route path=\\"/sicpjs/:section?\\">
      <SicpNavigationBar />
    </Route>
    <Route />
  </Switch>
</Fragment>"
`;
